[FLINK-9816][network] add option to configure SSL engine provider for TM communication

NicoK opened the pull request:
## What is the purpose of the change

Netty has the ability to run with different `SSLEngine` implementations but with our current setup, we are fixed to the JDK implementation, although one based on OpenSSL is expected to be faster [1].
We should make this configurable and ideally also provide everything needed to run with OpenSSL in the future (the last part is not part of this PR).

[1] https://netty.io/wiki/requirements-for-4.x.html#benefits-of-using-openssl

This PR subsumes #6328.

## Brief change log

- netty-fy SSL configuration in `SSLUtils` by using Netty's `SslContextBuilder` (only a few places do not use netty SSL setups - provide a workaround there)
- allow selecting the SSL engine provider via `security.ssl.provider`
- add openSSL-based SSL tests (if available) - some may currently fail due to different behaviour (this may need to be fixed once the second part is done)
- use `OpenSslX509KeyManagerFactory` for openSSL back-end

## Verifying this change

This change can be verified as follows:

- added unit tests
- ran e2e SSL test
- I verified by running an SSL setup with 2 TMs and submitting a job through the WebUI with the default `JDK` SSL engine and `OPENSSL`
-> all using a custom build using `netty-tcnative` with statically linked boringssl libraries from http://netty.io/wiki/forked-tomcat-native.html (see https://github.com/NicoK/flink-shaded/tree/flink-9816.static-6.0)

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? **docs, JavaDocs**

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java:
```diff
@@ -386,7 +386,8 @@ public void testCreateSSLEngineFactory() throws Exception {
 		final SSLHandlerFactory serverSSLHandlerFactory = SSLUtils.createInternalServerSSLEngineFactory(serverConfig);
-		final SslHandler sslHandler = serverSSLHandlerFactory.createNettySSLHandler();
+		// note: a 'null' allocator seems to work here (probably only because we do not use the ssl engine!)
+		final SslHandler sslHandler = serverSSLHandlerFactory.createNettySSLHandler(null);
```
`io.netty.buffer.UnpooledByteBufAllocator#DEFAULT`?

NicoK replied:
actually, a later commit in the PR already fixed that :)

pnowojski commented on flink-mesos/src/main/java/org/apache/flink/mesos/util/MesosArtifactServer.java:
```diff
@@ -135,7 +134,8 @@ protected void initChannel(SocketChannel ch) {
 				// SSL should be the first handler in the pipeline
 				if (sslFactory != null) {
-					ch.pipeline().addLast("ssl", sslFactory.createNettySSLHandler());
+					ch.pipeline().addLast("ssl",
```
I do not understand this `[FLINK-9816][network] netty-fy SSL configuration` commit. What does it do? There are no new tests, no changes in the existing tests, no documentation and nothing in the commit message :(

Is it refactor? If so please explain in the commit message what are you refactoring and why.

pnowojski commented on flink-core/src/main/java/org/apache/flink/configuration/SecurityOptions.java:
```diff
@@ -299,6 +300,26 @@
+					.list(
+						TextElement.text("%s: default Java-based SSL engine", TextElement.code("JDK")),
+						TextElement.text("%s: openSSL-based SSL engine using system libraries"
+							+ " (falls back to JDK if not available)", TextElement.code("OPENSSL"))
```
It would be safer to fail instead of fall back if users specifies the openSSL explicitly.

NicoK replied:
Do you think, it would then be a good idea to also have `AUTO` with the behaviour that it chooses openSSL if available, otherwise Java SSL?

NicoK replied:
It may be dangerous though if this would result in a heterogeneous cluster where some nodes had to fall back to JavaSSL while others are using openSSL. Maybe let's leave it with the user having to decide manually.

pnowojski replied:
That is a good question. I think an `AUTO` mode (even as the default value) would make sense. But I leave it up to you to decide whether this is worth the extra effort or not.

Problem in non heterogeneous shouldn't be that frequent. I was mainly concerned with the situation where somebody is explicitly setting the value to `OpenSSL` and it is silently not being respected.

NicoK replied:
I'll leave it without `AUTO` since, especially with standalone clusters, it may actually quite simple to accidentally get into a heterogeneous SSL setup by missing some dependencies on some nodes. I fear that debugging such scenarios would be painful

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/net/SSLUtils.java:
```diff
@@ -195,6 +198,22 @@ public static SSLHandlerFactory createRestClientSSLEngineFactory(final Configura
+		} else if (providerString.equalsIgnoreCase("JDK")) {
+			return JDK;
+		} else {
+			throw new IllegalArgumentException("Unknown SSL provider: " + providerString);
```
`IllegalArgumentException` runtime exception that doesn't use our exception hierarchy?  Did you mean `IllegalConfigurationException`?

NicoK replied:
sure, why not

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java:
```diff
@@ -52,6 +60,23 @@
+	public static final List<String> AVAILABLE_SSL_PROVIDERS;
+	static {
+		if (OpenSsl.isAvailable()) {
+			AVAILABLE_SSL_PROVIDERS = Arrays.asList("JDK", "OPENSSL");
```
Do we test for this somehow on travis? Manually? Is there some kind of instruction how to make `OpenSsl` available during testing?

NicoK replied:
For now, we'd have to have a custom flink-shaded build around*. There are licensing issues that prevent us from going the easy way and the actual way of providing openSSL in a safe way for the Flink distribution is not set in stone yet. I'll create a follow-up PR with proper documentation once this has been defined.

* I have two WIP branches for this: https://github.com/NicoK/flink-shaded/commits/flink-9816.static-6.0 and https://github.com/NicoK/flink-shaded/tree/flink-9816

NicoK replied:
I enriched this PR with a commit that enables the custom flink-shaded builds with static openSSL builds for Travis, just to see whether the tests run through. We would have to drop that commit during the merge though.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/NettyClientServerSslTest.java:
```diff
@@ -96,6 +108,7 @@ private void testValidSslConnection(Configuration sslConfig) throws Exception {
 		// session context is only be available after a session was setup -> this should be true after data was sent
 		SSLSessionContext sessionContext = sslHandler.engine().getSession().getSessionContext();
 		assertNotNull("bug in unit test setup: session context not available", sessionContext);
+		// TODO: sessionContext is from the client side which may not have a session cache at all (with openSSL it behaves that way)
```
?

NicoK replied:
done - by a little extra work here and there :(

NicoK commented:
Thanks for the review, I addressed all issues you found and rebased onto the latest `master` so we don't have conflicts and non-related commits anymore (except for one testing enrichment marked with `DO-NOT-MERGE`)

NicoK commented:
hi @pnowojski I did some updates to this PR:
- rebased onto latest master
- changed unit tests to work with the latest flink-shaded openSSL jars
- set nightly e2e tests to use both, dynamically and statically linked openSSL variants for the `test_streaming_file_sink`
- updated docs
- a few hotfixes
-> everything starting b21a1d0296c66facae67cb5831350f224c3afc5f are new commits

pnowojski commented on flink-end-to-end-tests/test-scripts/test_streaming_file_sink.sh:
```diff
@@ -22,8 +22,11 @@ OUT_TYPE="${1:-local}" # other type: s3
 source "$(dirname "$0")"/common_s3.sh
 
+# randomly set up openSSL with dynamically/statically linked libraries
+OPENSSL_LINKAGE=$(if (( RANDOM % 2 )) ; then echo "dynamic"; else echo "static"; fi)
```
add an echo message directly after the `RANDOM` what was chosen to be tested

NicoK replied:
don't you think it is enough to print this one (during setup, in common_ssl.sh)?
```echo "Setting up SSL with: ${type} ${provider} ${provider_lib}"```

pnowojski replied:
Currently yes, but it would be safer to tie those two things together (best hidden behind some function:
```
int x = new Random().nextInt();
System.out.println(format("Executing test with following value of x = [%s] that was randomly selected", x));
```
), so that there is obvious connection between randomness and printing the random result.

Otherwise, someone might remove the logging message, without realising the implication of that fact.

NicoK replied:
I'm not entirely convinced but also don't object and therefore added a fixup
-> I'm now waiting for the flink-shaded 7.0 release before merging
